contributors/design-proposals/sig-cli/kubectl_plan.md

```diff
@@ -0,0 +1,164 @@
+# Kubectl Plan
+
+## TL;DR
+
+Introduce a new `plan` command similar to `kubectl create --dry-run -o yaml`
```

Comment:
I'm not a fan of a new command group.

Please see my comment here:
https://github.com/kubernetes/kubernetes/issues/30558#issuecomment-331342865

I am in favor of simplifying the flags and cleaning up the output. 

Also, I don't find the term `plan` to be intuitive. Is that borrowed from Terraform?

Reply:
Yeah it was suggested by @jbeda for that reason.  I am not tied to the name.

I am interested to get your feedback on adding support for implementing the `create` subcommands as sub resources on the server side, and dynamically discovering them and their arguments through the discovery service + openapi extensions.

Reply:
Re. create: I've thought about that in the past:
https://github.com/kubernetes/kubernetes/issues/5280

I have significant reservations, though. We've added a number of options to these generations. As with template parameterization, there is a slippery slope where we just develop an alternative API schema that's almost as complex but different than the current one.

My instinct is to find another way to do this.

What's your main reason for wanting a server-side implementation?

Reply:
> What's your main reason for wanting a server-side implementation?

There are 2 problems I would like to solve and think dynamically discovered server-side commands would provide a solution:

1) Tie the resource-specific commands to the resource APIs actually available in the cluster.

- In a versioned skewed environment, as a kubectl user, I would like kubectl to show and support the resource specific commands/subcommands for the APIs present in the cluster, rather than those compiled into the client.
- When using extension APIs, as a kubectl user, I would like to be able to view and use resource specific commands/subcommands of the extension APIs in the cluster without separately installing a compiled plugin (e.g. simplifies questions such as: what if the extension is in some cluster but not others?  or what if my corporate IT policy doesn't allow me to run arbitrary binaries on my computer?).

2) Allow end-to-end development of APIs to be performed without changing kubectl code for things like the `create resource name`

Currently, adding a new API to Kubernetes results in new specific kubectl commands commands added for that resource.  The code for these commands lives in the kubectl code base, but the ownership of their maintenance is unclear (is it the maintainer of the API code, or the maintainer of the kubectl codebase).  We frequently discover issues in kubectl API-specific commands late into the release cycle, and resolving these issues generally requires a fire drill from both an API maintainer and an kubectl maintainer.

Reply:
Version skew, extensibility, and maintainability -- reasonable objectives.

Quick thought: How would users do this for CRDs?

create configmap and secret are somewhat special, since they transform non-API data in support of particular idioms/patterns. We do need an API-like representation for the declarative manifest work, but there's still client-side logic that is required, such as reading and combining multiple files.

run and expose are also somewhat special, since we've added a fairly large number of options to run in support of common use cases, they span multiple resource types, the commands have existed for a long time, and they target Kubernetes's most central user-facing functionality, which has also existed for a fairly long time. 

I'd argue that we should put more than typical effort into polished, opinionated/curated experiences for those 4 commands in particular, even if somewhat ad hoc, in support of very common use cases and reducing the learning curve.

Other current commands are more basic, narrow, and straightforward: clusterrole, deployment, namespace, pdb, quota, role, rolebinding, service, serviceaccount. These are more targeted towards generating boilerplate and scaffolding the API schema structure, to reduce work, the learning curve, and errors. These seem more amenable to more scalable approaches. My concerns about growing alternative schemas stand, however.

The alternatives don't appear to have been thoroughly vetted. 

Have you seen codesolvent.com, yipee.io, or the vscode-based autocompletion? Maybe we could attach enough information to OpenAPI to build a generic resource attribute discovery wizard, even as a CLI or shell environment. 

Or, we could use it to generate a resource with all default values, include fake names for things like images, which the user could then just edit, perhaps using kubectl set(*).

Or, we could hand-write examples with all default values and populate the example data in OpenAPI: https://swagger.io/docs/specification/adding-examples/

I kind of like that last one, actually. It would improve our generated API documentation, also.

I also want a generic search/browse mechanism that would allow us to build a catalog of more specific examples (e.g., Deployment for a java app), but that's a separate problem.

(*) Does `kubectl set` suffer from similar problems, or is it mainly generators that are painful? I do want `set` to work in the client. I'm hoping that by targeting super-common, stable use cases (e.g., set image) it will be easier to maintain.

Reply:
Note that another advantage of example text is that it can include comments (assuming yaml).

Reply:
In terms of how to request an example, this is also sort of related to #123.

Reply:
> but there's still client-side logic that is required, such as reading and combining multiple files.

This can still be done dynamically

> run and expose are also somewhat special, since we've added a fairly large number of options to run in support of common use cases, they span multiple resource types, the commands have existed for a long time, and they target Kubernetes's most central user-facing functionality, which has also existed for a fairly long time.

I think these commands cover too many use cases and can be a bit of a crutch.  Folks have asked for more flags to be added to run for more special cases.  I'd rather marshal our efforts around commands to manage config, and provide a few transformations.